redirector.rs/src/main.rs

```diff
@@ -0,0 +1,179 @@
+    }
+
+    fn link_attr(&self, ino: u64) -> FileAttr {
+        FileAttr {
+            ino,
+            size: 0,
+            blocks: 0,
+            atime: self.start_time,
+            mtime: self.start_time,
+            ctime: self.start_time,
+            crtime: self.start_time,
+            kind: FileType::Symlink,
+            perm: 0o644,
+            nlink: 0,
+            uid: 0,
+            gid: 0,
+            rdev: 0,
+            flags: 0,
+        }
+    }
+}
+
+impl Filesystem for Redirector {
+    fn lookup(&mut self, _req: &Request, _parent: u64, name: &OsStr, reply: ReplyEntry) {
+        match name.to_str() {
+            Some("private") => reply.entry(&TTL, &self.link_attr(PRIVATE_LINK_INODE), 0),
+            Some("public") => reply.entry(&TTL, &self.link_attr(PUBLIC_LINK_INODE), 0),
+            Some("team") => reply.entry(&TTL, &self.link_attr(TEAM_LINK_INODE), 0),
+            _ => reply.error(ENOENT),
+        }
+    }
+
+    fn getattr(&mut self, _req: &Request, ino: u64, reply: ReplyAttr) {
+        match ino {
+            ROOT_DIR_INODE => reply.attr(&TTL, &self.root_dir_attr()),
+            PRIVATE_LINK_INODE | PUBLIC_LINK_INODE | TEAM_LINK_INODE => {
+                reply.attr(&TTL, &self.link_attr(ino))
+            }
+            _ => reply.error(ENOENT),
+        }
+    }
+
+    fn readlink(&mut self, req: &Request, ino: u64, reply: ReplyData) {
+        let subdir = match ino {
+            PRIVATE_LINK_INODE => "private",
+            PUBLIC_LINK_INODE => "public",
+            TEAM_LINK_INODE => "team",
+            _ => return reply.error(ENOENT),
+        };
+        // TODO: Should we cache the mountpoint somehow?
+        if let Some(mountpoint) = get_keybase_fuse_mount(req.uid()) {
+            let target = mountpoint.join(subdir);
+            reply.data(target.as_os_str().as_bytes());
+        } else {
+            reply.error(ENOENT);
+        }
+    }
+
+    fn readdir(
+        &mut self,
+        _req: &Request,
+        ino: u64,
+        _fh: u64,
+        offset: i64,
+        mut reply: ReplyDirectory,
+    ) {
+        if ino == ROOT_DIR_INODE {
+            if offset == 0 {
+                reply.add(ROOT_DIR_INODE, 0, FileType::Directory, ".");
+                reply.add(ROOT_DIR_INODE, 1, FileType::Directory, "..");
+                reply.add(PRIVATE_LINK_INODE, 2, FileType::Symlink, "private");
+                reply.add(PUBLIC_LINK_INODE, 3, FileType::Symlink, "public");
+                reply.add(TEAM_LINK_INODE, 4, FileType::Symlink, "team");
+            }
+            reply.ok();
+        } else {
+            reply.error(ENOENT);
+        }
+    }
+}
+
+// This is explicitly only implemented for Linux, so that building on other
+// platforms is a compiler error.
+#[cfg(target_os = "linux")]
+fn get_keybase_fuse_mount(uid: u32) -> Option<PathBuf> {
+    let uid_str = format!("user_id={}", uid);
+    let file = File::open("/proc/mounts").expect("failed to open /proc/mounts");
```

Comment:
I wonder if it's better to call out to `setmntent`. Or at least look at its source to make sure we're mimicking its behavior.